src/game/puppyprint.c

```diff
@@ -1003,68 +1004,97 @@ void print_small_text(s32 x, s32 y, const char *str, s32 align, s32 amount, s32
-            s32 g = (((str[i +  7] >= 'A') ? str[i +  7] - 'A' + 0xA : str[i +  7] - '0') * 16) + ((str[i +  8] >= 'A') ? str[i +  8] - 'A' + 0xA : str[i +  8] - '0');
-            s32 b = (((str[i +  9] >= 'A') ? str[i +  9] - 'A' + 0xA : str[i +  9] - '0') * 16) + ((str[i +  10] >= 'A') ? str[i +  10] - 'A' + 0xA : str[i +  10] - '0');
-            s32 a = (((str[i +  11] >= 'A') ? str[i +  11] - 'A' + 0xA : str[i +  11] - '0') * 16) + ((str[i +  12] >= 'A') ? str[i +  12] - 'A' + 0xA : str[i +  12] - '0');
-            print_set_envcolour(r, g, b, a);
-        } else if (strncmp((str + i), "<FADE_xxxxxxxx,xxxxxxxx,xx>", 6) == 0) { // Same as above, except it fades between two colours. The third set of numbers is the speed it fades.
-            s32 r = (((str[i +  6] >= 'A') ? str[i +  6] - 'A' + 0xA : str[i +  6] - '0') * 16) + ((str[i +  7] >= 'A') ? str[i +  7] - 'A' + 0xA : str[i +  7] - '0');
-            s32 g = (((str[i +  9] >= 'A') ? str[i +  9] - 'A' + 0xA : str[i +  9] - '0') * 16) + ((str[i +90] >= 'A') ? str[i +  9] - 'A' + 0xA : str[i +  9] - '0');
-            s32 b = (((str[i +  10] >= 'A') ? str[i +  10] - 'A' + 0xA : str[i +  10] - '0') * 16) + ((str[i +  11] >= 'A') ? str[i +  11] - 'A' + 0xA : str[i +  11] - '0');
-            s32 a = (((str[i +  12] >= 'A') ? str[i +  12] - 'A' + 0xA : str[i +  12] - '0') * 16) + ((str[i +  13] >= 'A') ? str[i +  13] - 'A' + 0xA : str[i +  13] - '0');
-            s32 r2 = (((str[i +  15] >= 'A') ? str[i +  15] - 'A' + 0xA : str[i +  15] - '0') * 16) + ((str[i +  16] >= 'A') ? str[i +  16] - 'A' + 0xA : str[i +  16] - '0');
-            s32 g2 = (((str[i +  17] >= 'A') ? str[i +  17] - 'A' + 0xA : str[i +  17] - '0') * 16) + ((str[i +  18] >= 'A') ? str[i +  18] - 'A' + 0xA : str[i +  18] - '0');
-            s32 b2 = (((str[i +  19] >= 'A') ? str[i +  19] - 'A' + 0xA : str[i +  19] - '0') * 16) + ((str[i +  20] >= 'A') ? str[i +  20] - 'A' + 0xA : str[i +  20] - '0');
-            s32 a2 = (((str[i +  21] >= 'A') ? str[i +  21] - 'A' + 0xA : str[i +  21] - '0') * 16) + ((str[i +  22] >= 'A') ? str[i +  22] - 'A' + 0xA : str[i +  22] - '0');
-            s32 spd = (((str[i +  24] >= 'A') ? str[i +  24] - 'A' + 0xA : str[i +  24] - '0') * 16) + ((str[i +  25] >= 'A') ? str[i +  25] - 'A' + 0xA : str[i +  25] - '0');
-            // Find the median.
-            s32 r3 = (r + r2) / 2;
-            s32 g3 = (g + g2) / 2;
-            s32 b3 = (b + b2) / 2;
-            s32 a3 = (a + a2) / 2;
-            // Find the difference.
-            s32 r4 = (r - r2) / 2;
-            s32 g4 = (g - g2) / 2;
-            s32 b4 = (b - b2) / 2;
-            s32 a4 = (a - a2) / 2;
-            // Now start from the median, and wave from end to end with the difference, to create the fading effect.
+        if (strncmp((newStr), "<COL_xxxxxxxx>", 5) == 0) { // Simple text colour effect. goes up to FF for each, so FF0000FF is red.
```

Comment:
`len` check should probably be added inside this block to help evade potential UB.